src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -131,32 +131,60 @@ DropMarkedShards(bool waitForCleanupLock)
 			continue;
 		}
 
-		ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
-															placement->placementId);
-		ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
-
-		ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
-							 INT64_FORMAT " on %s:%d after it was moved away",
-							 shardPlacement->placementId, shardPlacement->shardId,
-							 shardPlacement->nodeName, shardPlacement->nodePort)));
-
-		/* prepare sql query to execute to drop the shard */
-		StringInfo dropQuery = makeStringInfo();
-		char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
-		appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+		if (TryDropShard(placement))
+		{
+			removedShardCount++;
+		}
+	}
 
-		List *dropCommandList = list_make2("SET LOCAL lock_timeout TO '1s'",
-										   dropQuery->data);
+	return removedShardCount;
+}
 
-		/* remove the shard from the node and the placement information */
-		SendCommandListToWorkerInSingleTransaction(shardPlacement->nodeName,
-												   shardPlacement->nodePort,
-												   NULL, dropCommandList);
 
+/*
+ * TryDropShard tries to drop the given shard placement and returns
+ * true on success. On failure, this method swallows errors and emits them
+ * as WARNINGs.
+ */
+static bool
+TryDropShard(GroupShardPlacement *placement)
+{
+	ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
+														placement->placementId);
+	ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
+
+	ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
+						 INT64_FORMAT " on %s:%d after it was moved away",
+						 shardPlacement->placementId, shardPlacement->shardId,
+						 shardPlacement->nodeName, shardPlacement->nodePort)));
+
+	/* prepare sql query to execute to drop the shard */
+	StringInfo dropQuery = makeStringInfo();
+	char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
+	appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+
+	/*
+	 * We set a lock_timeout here so that if there are running queries on the
+	 * shards we won't get blocked more than 1s and fail.
+	 *
+	 * The lock timeout also avoids getting stuck in a distributed deadlock, which
+	 * can occur because we might be holding pg_dist_placement locks while also
+	 * taking locks on the shard placements, and this code interrupts the
+	 * distributed deadlock detector.
+	 */
+	List *dropCommandList = list_make2("SET LOCAL lock_timeout TO '1s'",
+									   dropQuery->data);
+
+	/* remove the shard from the node */
+	bool success =
+		SendOptionalCommandListToWorkerInTransaction(shardPlacement->nodeName,
+													 shardPlacement->nodePort,
+													 NULL, dropCommandList);
+	if (success)
+	{
+		/* delete the actual placement */
 		DeleteShardPlacementRow(placement->placementId);
-
-		removedShardCount++;
 	}
```

Comment:
should we have some kind of warning here to indicate that dropping failed?

Reply:
umm, a warning could be useful. Printing every failure could result in cluttering in the logs under some situations, so what about we warn at the end only once saying how many drop ops failed?

Reply:
How does this look @marcocitus https://github.com/citusdata/citus/pull/4919/files#diff-dae583fd6d707aa7fd8ec0849dc8462a0766446152f84cfb7a56698edb1fae67R147?

I will try to add a test to cover this as well

Reply:
I have added a test